File: frontend/src/api/modelCenter.ts
import apiClient from '@/api/client';

export interface FeatureCoverageItem {
  feature_name: string;
  feature_label: string;
  available_count: number;
  missing_count: number;
  coverage_rate: number;
  variance: number;
  selected: boolean;
}

export interface ValueCountItem {
  name: string;
  value: number;
}

export interface OutcomeSummary {
  min_value: number;
  max_value: number;
  mean_value: number;
}

export interface DatasetOverviewResponse {
  total_patients: number;
  eligible_records: number;
  dropped_records: number;
  selected_feature_names: string[];
  treatment_name: string;
  control_name: string;
  outcome_name: string;
  treatment_distribution: ValueCountItem[];
  outcome_summary: OutcomeSummary | null;
  feature_coverage: FeatureCoverageItem[];
  dropped_examples: string[];
  assumptions: string[];
  limitations: string[];
}

export interface CausalTrainingRequest {
  model_name?: string;
  test_ratio: number;
  random_seed: number;
  max_features: number;
  min_feature_coverage: number;
  feature_names?: string[];
  activate_after_train: boolean;
}

export interface ModelVersionSummary {
  id: number;
  name: string;
  version_type: string;
  status: string;
  description: string | null;
  artifact_path: string | null;
  engine_backend: string | null;
  estimator_message: string | null;
  reproducibility_status: string | null;
  metrics: Record<string, unknown>;
  config: Record<string, unknown>;
  feature_list: string[];
  selected_feature_keys: string[];
  random_seed: number | null;
  test_ratio: number | null;
  min_feature_coverage: number | null;
  artifact_generated_at: string | null;
  training_started_at: string | null;
  training_completed_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface ActiveModelResponse {
  active_model: ModelVersionSummary | null;
}

export interface HistogramBucket {
  name: string;
  value: number;
}

export interface FeatureImportanceItem {
  feature_name: string;
  feature_label: string;
  importance: number;
}

export interface SubgroupResultItem {
  feature_name: string;
  feature_label: string;
  subgroup_name: string;
  sample_count: number;
  average_ite: number;
}

export interface PatientEffectItem {
  patient_id: number;
  patient_code: string;
  anonymized_code: string;
  treatment_label: string;
  observed_outcome: number;
  estimated_ite: number;
}

export interface CausalEvaluationResultResponse {
  model_version: ModelVersionSummary;
  ate: number;
  validation_ate: number | null;
  observed_group_difference: number | null;
  engine_backend: string;
  estimator_message: string;
  reproducibility_status: string;
  dataset_record_count: number;
  train_record_count: number;
  validation_record_count: number;
  treatment_name: string;
  control_name: string;
  outcome_name: string;
  selected_feature_names: string[];
  selected_feature_keys: string[];
  random_seed: number | null;
  test_ratio: number | null;
  min_feature_coverage: number | null;
  artifact_generated_at: string | null;
  ite_distribution: HistogramBucket[];
  feature_importance: FeatureImportanceItem[];
  subgroup_results: SubgroupResultItem[];
  top_positive_patients: PatientEffectItem[];
  top_negative_patients: PatientEffectItem[];
  assumptions: string[];
  limitations: string[];
}

export interface CausalTrainingResponse {
  message: string;
  model_version: ModelVersionSummary;
  result: CausalEvaluationResultResponse;
}

export async function fetchCausalDatasetOverview(params?: {
  max_features?: number;
  min_feature_coverage?: number;
}) {
  const { data } = await apiClient.get<DatasetOverviewResponse>('/model-center/dataset-overview', {
    params,
  });
  return data;
}

export async function trainCausalModel(payload: CausalTrainingRequest) {
  const { data } = await apiClient.post<CausalTrainingResponse>('/model-center/train', payload);
  return data;
}

export async function fetchModelVersions(versionType?: string) {
  const { data } = await apiClient.get<ModelVersionSummary[]>('/model-center/versions', {
    params: {
      version_type: versionType,
    },
  });
  return data;
}

export async function fetchActiveModel(versionType = 'causal') {
  const { data } = await apiClient.get<ActiveModelResponse>('/model-center/active-model', {
    params: {
      version_type: versionType,
    },
  });
  return data;
}

export async function activateModelVersion(versionId: number) {
  const { data } = await apiClient.post<ModelVersionSummary>(`/model-center/versions/${versionId}/activate`);
  return data;
}

export async function fetchCausalResults(modelVersionId?: number) {
  const { data } = await apiClient.get<CausalEvaluationResultResponse>('/model-center/causal-results', {
    params: {
      model_version_id: modelVersionId,
    },
  });
  return data;
}

File: backend/app/schemas/model_center.py
from datetime import datetime

from pydantic import BaseModel, Field


class FeatureCoverageItem(BaseModel):
    feature_name: str
    feature_label: str
    available_count: int
    missing_count: int
    coverage_rate: float
    variance: float
    selected: bool


class ValueCountItem(BaseModel):
    name: str
    value: int


class OutcomeSummary(BaseModel):
    min_value: float
    max_value: float
    mean_value: float


class DatasetOverviewResponse(BaseModel):
    total_patients: int
    eligible_records: int
    dropped_records: int
    selected_feature_names: list[str]
    treatment_name: str
    control_name: str
    outcome_name: str
    treatment_distribution: list[ValueCountItem]
    outcome_summary: OutcomeSummary | None = None
    feature_coverage: list[FeatureCoverageItem]
    dropped_examples: list[str]
    assumptions: list[str]
    limitations: list[str]


class CausalTrainingRequest(BaseModel):
    model_name: str | None = None
    test_ratio: float = Field(default=0.2, gt=0, lt=0.5)
    random_seed: int = Field(default=20260319)
    max_features: int = Field(default=10, ge=4, le=16)
    min_feature_coverage: float = Field(default=0.7, ge=0.3, le=1.0)
    feature_names: list[str] | None = None
    activate_after_train: bool = True


class ModelVersionSummary(BaseModel):
    id: int
    name: str
    version_type: str
    status: str
    description: str | None = None
    artifact_path: str | None = None
    engine_backend: str | None = None
    estimator_message: str | None = None
    reproducibility_status: str | None = None
    metrics: dict
    config: dict
    feature_list: list[str]
    selected_feature_keys: list[str] = []
    random_seed: int | None = None
    test_ratio: float | None = None
    min_feature_coverage: float | None = None
    artifact_generated_at: datetime | None = None
    training_started_at: datetime | None = None
    training_completed_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class ActiveModelResponse(BaseModel):
    active_model: ModelVersionSummary | None = None


class HistogramBucket(BaseModel):
    name: str
    value: int


class FeatureImportanceItem(BaseModel):
    feature_name: str
    feature_label: str
    importance: float


class SubgroupResultItem(BaseModel):
    feature_name: str
    feature_label: str
    subgroup_name: str
    sample_count: int
    average_ite: float


class PatientEffectItem(BaseModel):
    patient_id: int
    patient_code: str
    anonymized_code: str
    treatment_label: str
    observed_outcome: float
    estimated_ite: float


class CausalEvaluationResultResponse(BaseModel):
    model_version: ModelVersionSummary
    ate: float
    validation_ate: float | None = None
    observed_group_difference: float | None = None
    engine_backend: str
    estimator_message: str
    reproducibility_status: str
    dataset_record_count: int
    train_record_count: int
    validation_record_count: int
    treatment_name: str
    control_name: str
    outcome_name: str
    selected_feature_names: list[str]
    selected_feature_keys: list[str]
    random_seed: int | None = None
    test_ratio: float | None = None
    min_feature_coverage: float | None = None
    artifact_generated_at: datetime | None = None
    ite_distribution: list[HistogramBucket]
    feature_importance: list[FeatureImportanceItem]
    subgroup_results: list[SubgroupResultItem]
    top_positive_patients: list[PatientEffectItem]
    top_negative_patients: list[PatientEffectItem]
    assumptions: list[str]
    limitations: list[str]


class CausalTrainingResponse(BaseModel):
    message: str
    model_version: ModelVersionSummary
    result: CausalEvaluationResultResponse
